### PR TITLE
Check whether the file id is valid, before using it to delete the previews

### LINKED ISSUE #14671: Thumbnail cache gets deleted after upload from smartphone
Steps to reproduce

Take a picture with your smartphone (Instant upload enabled, filesize >1MB)
check your thumbnails cache directory on the server ($OC_HOME/data/user/thumbnails)

Expected result

Thumbnail cache persists upload

Actual result

The complete directory $OC_HOME/data/user/thumbnails gets deleted

Versions

ownCloud 8 (stable8)

Hints

I've traced the deletion of the folder back to this function call in $OC_HOME/lib/private/preview.php:

```php
/**
 * deletes all previews of a file
 */
public function deleteAllPreviews() {
        $file = $this->getFile();

        $fileInfo = $this->getFileInfo($file);

        $toDelete = $this->getChildren();
        $toDelete[] = $fileInfo;

        foreach ($toDelete as $delete) {
                if ($delete !== null && $delete !== false) {
                        /** @var \OCP\Files\FileInfo $delete */
                        $fileId = $delete->getId();

                        $previewPath = $this->getPreviewPath($fileId);
                        $this->userView->deleteAll($previewPath);
                        $this->userView->rmdir($previewPath);
                }
        }
}
```

If $fileId is null (for whatever reason) $previewPath results in thumbnails// and rmdir removes the whole directory.

This happens on my server every time i upload a picture and it's pretty annoying :-\

### PR DESCRIPTION
**Pull request by nickvergessen**
Fix #14671 

@oparoz @jmit79 @LukasReschke @MorrisJobke

### CODE REVIEW

**Comment by LukasReschke** [comment]
:+1: – nasty bug…

File: lib/private/preview.php [hunk 1]
```diff
@@ -339,9 +337,13 @@ public function deleteAllPreviews() {
 				/** @var \OCP\Files\FileInfo $delete */
 				$fileId = $delete->getId();
 
-				$previewPath = $this->getPreviewPath($fileId);
-				$this->userView->deleteAll($previewPath);
-				$this->userView->rmdir($previewPath);
+				// getId() might return null, e.g. when the file is a
+				// .ocTransferId*.part file from chunked file upload.
+				if ($fileId !== null) {
```
Comment (oparoz) [inline lib/private/preview.php]: Maybe use empty() to cover all fail cases?

Comment (nickvergessen) [inline lib/private/preview.php]: or just `if ($fileId)` I think all 3 versions work.

Comment (oparoz) [inline lib/private/preview.php]: `!== null` does not cover `[]` or `false`, but `($fileId)` would work :)
http://php.net/manual/en/types.comparisons.php

Comment (nickvergessen) [inline lib/private/preview.php]: Yes, I know, but getId can not return those :P

Comment (nickvergessen) [inline lib/private/preview.php]: changed to `!empty()`

**Comment by oparoz** [comment]
OK, so I was distracted earlier and missed that the check was on the ID... 

What we should probably do though (as a bonus while we're here) is add a check to `$fileInfo = $this->getFileInfo();` because that's where we can get `[]` or `null` and any method call on `$fileInfo` will generate a fatal error. 
ON 8.0.1, `getFileInfo()` throws a `NotFoundException`

**Comment by MorrisJobke** [comment]
@nickvergessen Can you have a look at @oparoz comment?

**Comment by nickvergessen** [comment]
Updated

**Comment by oparoz** [comment]
:+1: Well done! :)

**Comment by MorrisJobke** [comment]
@nickvergessen What about a backport?

**Comment by nickvergessen** [comment]
Will prepare soon, but not before 8.0.2 anyway

**Pull request merged**